VSR: Batch pending client requests

kprotty opened the pull request:
Implements #489

kprotty committed:
vsr: client batching

Removes vsr.Client's `request()` in favor of the batch API as described
in #489. Components which need to work with Messages directly
must use `raw_request()`.

batiati reviewed (commented):
TIL `switch inline x,y,z` was possible⚡

kprotty committed:
client: fix EchoClient and various typos

kprotty committed:
benchmark: fix vsr.Client typo

batiati commented on src/testing/state_machine.zig:
```diff
@@ -22,6 +22,23 @@ pub fn StateMachineType(
+            demuxed: []u8,
+            event: struct { index: u32, count: u32 },
+        ) u32 {
+            _ = .{ operation, results, demuxed, event };
```
Nice pattern 👍

batiati commented on src/vsr/client.zig:
```diff
@@ -26,12 +28,25 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+            demux_queue: FIFO(Demux) = .{ .name = null },
         };
 
+        const DemuxPool = IOPS(Request.Demux, constants.client_request_queue_max);
```
Not sure if we want to use `client_request_queue_max` here.
Theoretically, we could have 8k single-item batches per request * `client_request_queue_max`, which is a lot to preallocate.
That's why we need a `DemuxPool` and a new limit `batch_logical_max`.

We can receive `batch_logical_max` as a comptime parameter in `vsr.Client`, allowing the user to set it to `0` if they want to disable batching completely.

batiati commented on src/vsr/client.zig:
```diff
@@ -196,50 +226,142 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+                const body = batch.request.message.body();
+                return body[body_offset..];
+            }
+        };
```
Let's move this struct to the top, among the other decls.

batiati commented on src/vsr/client.zig:
```diff
@@ -196,50 +226,142 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
-        ) void {
+            event_count: usize,
+        ) BatchError!Batch {
             assert(@intFromEnum(operation) >= constants.vsr_operations_reserved);
```
Can you document here that a `Batch` cannot be stored, and once acquired it cannot be released back and the user must call `batch_submit` before the next `tick()`?

It's a quite sensitive API change.

batiati commented on src/vsr/client.zig:
```diff
@@ -196,50 +226,142 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+                    if (request.message.header.size + body_size > constants.message_size_max) continue;
+
+                    // Reserve a Demux node for attaching this Batch onto the existing Request.
+                    // TODO: Should this return TooManyOutstanding? Currently, it falls back to new.
```
It should fallback to a new request, which can return `TooManyOutstanding` if there's no message available.

batiati commented on src/vsr/client.zig:
```diff
@@ -196,50 +226,142 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+                            }
+                            const request_body_size = request.message.header.size - @sizeOf(Header);
+                            break :blk @intCast(@divExact(request_body_size, event_size));
+                        },
```
- Use `else`
- Can we use `request.message.body().len` instead of recalculating it?

batiati commented on src/state_machine.zig:
```diff
@@ -100,6 +100,47 @@ pub fn StateMachineType(
+        ) u32 {
+            switch (operation) {
+                inline .create_accounts, .create_transfers => |op| {
+                    const out_results = std.mem.bytesAsSlice(Result(op), demuxed);
```
add `assert(batch_logical(op))`

batiati commented on src/vsr/client.zig:
```diff
@@ -78,10 +93,20 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+
+        /// StateMachine can't take advantage of Request.demux_queue order for in-place demuxing,
+        /// so instead it demuxes into a scratch buffer and that's passed into Request.Callback.
+        demux_buffer: *[constants.message_body_size_max]u8,
```
This buffer makes the support for batching generic for any state machine implementation, but it comes with the cost of one more copy ...

Since the Ops `create_accounts` and `create_transfers` can be demuxed in place, I think it's ok to assume that any state machine that supports batches also needs a way to demux in place.

Let's remove this buffer and if it's going to be a problem in the future we can always reconsider and introduce it back.

batiati commented on src/state_machine.zig:
Could you add unit tests for all demux/batching logic?

kprotty committed:
vsr: rework client demux

* Removes the demux_buffer in favor of an in-place StateMachine.Demuxer
* Clarifies comments about using the client Batch struct

kprotty committed:
vsr: client batching unit test

batiati commented on src/vsr/client.zig:
```diff
@@ -26,10 +28,83 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+                demux.next = pool.free;
+                pool.free = demux;
+            }
         };
```
Pretty cool, lazily populating the `free` list during `release` and not during the initialization 🤓🙌
Just one thing feels strange: `pool.used` shouldn't be named `used` since it's not decremented during `release`.  It's more like "the number of items that now live in the linked list". Can you think of something better?

On the other hand, we are adding an unnecessary `if` branch in the hot path `acquire` instead of populating the linked list during `init` time. Wdyt?

kprotty replied:
What about `pool.memory_allocated`?

As for the `memory.len` check, it happened lazily to avoid faulting/iterating 12MB worth of tiny items on init. It's always predicted false (excluding the odd worst-case scenario down below) so it's similar to `+` operator which does overflow checks in ReleaseSafe.

batiati commented on src/vsr/client.zig:
```diff
@@ -196,50 +282,117 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+        /// The returned Batch provides a slice of memory to write the events to. Once populated,
+        /// use `batch_submit` to commit the writes to the Message and send it through the Client.
+        /// NOTE: A Batch cannot be stored and must be passed to batch_submit() before any calling
+        /// any other methods on the Client.
```
```suggestion
        /// NOTE: A Batch cannot be stored and must be passed to batch_submit() before calling
        /// any other methods on the Client.
```

batiati commented on src/state_machine.zig:
```diff
@@ -42,6 +42,18 @@ pub fn StateMachineType(
+                }
+                break :blk max * global_constants.client_request_queue_max;
+            };
+
```
The `batch_logical_max` calculation is correct. However, we might not want to allocate that much memory (about 12MB, right?), given it's an odd worst-case scenario.

I suggest we keep it configurable per `vsr.Client`, since the caller site knows better the expected batching usage (eg. anything from `0` to `batch_logical_max` inclusive). 
For our `tb_client`, we can define an arbitrary default value not greater than a few thousand batches.

sentientwaffle replied:
I think `batch_logical_max` doesn't belong in the `StateMachine` -- it should be in `Client` instead, if it is needed.

The algorithm to compute `batch_logical_max` would be identical across any state machine implementation. We want to minimize the surface area for implementing new state machines.

> I suggest we keep it configurable per vsr.Client,

That is an improvement, but why not go with a context-style API:
- Then we can get rid of `batch_logical_max` and `client_request_queue_max`? (The limit moves up a layer. Clients could have no limit (by using dynamic allocator), or their own static limit.)
- Reduced `Client` surface area: Back to a single `request` function, instead of both `batch_get` and `batch_submit`.
- Reduced `Client` surface area: I think there would be no possible `error` on the request method.

batiati replied:
> That is an improvement, but why not go with a context-style API:
>
> - Then we can get rid of batch_logical_max and client_request_queue_max? (The limit moves up a layer. Clients could have no limit (by using dynamic allocator), or their own static limit.)
> - Reduced Client surface area: Back to a single request function, instead of both batch_get and batch_submit.
> - Reduced Client surface area: I think there would be no possible error on the request method.

I agree we can improve the client API, however, I thought it was out of the scope of this PR.
Can you elaborate on the suggestion of a context-style API a bit further?

...

The current implementation of the `tb_client` can take advantage of 4096 batches at most since it's the hard-coded upper `concurrency_max` limit.
That's why I think it makes sense for the caller to save memory specifying a new limit, lower than `batch_logical_max`.

sentientwaffle replied:
> I agree we can improve the client API, however, I thought it was out of the scope of this PR.

Sounds good to me.

batiati reviewed (changes requested):
The unit test for the client + message bus is great! Can we add units at the state machine as well to test the specific demux logic?

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -100,6 +112,51 @@ pub fn StateMachineType(
+                .create_accounts, .create_transfers => true,
+                .lookup_accounts, .lookup_transfers => false, // Don't batch lookups for now.
+            };
+        }
```
Rather than a function, this should be a `std.enums.EnumArray(Operation, bool)`.
That encodes the intent more directly -- this isn't a runtime decision, it shouldn't every vary.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -100,6 +112,51 @@ pub fn StateMachineType(
+                results: []CreateResult,
+
+                /// Create a Demuxer which can extract Results out of the reply bytes in-place.
+                pub fn init(reply: []align(@alignOf(CreateResult)) u8) Demuxer {
```
Why not take `[]CreateResult` as an argument?

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -100,6 +112,51 @@ pub fn StateMachineType(
+                    self: *Demuxer,
+                    event_offset: u32,
+                    event_size: u32,
+                ) []align(@alignOf(CreateResult)) u8 {
```
Maybe return `[]CreateResult` (or `[]const CreateResult`, if that would work), and let the caller cast to bytes? (To move logic out of the state machine, and make the type signature reflect more closely the actual data.)

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -26,10 +28,83 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+
+        pub const Batch = struct {
+            request: *Request,
+            demux: ?*Request.Demux, // Null iff batch_get() newly created `request`
```
Comment seems ungrammatical, or at least I can't parse it. Also missing punctuation.
Does it mean "Null iff batch_get() **is a** newly created `request`."?

kprotty replied:
"If `demux` is null, the `request` is newly create."

"If `demux` is _not_ null, the `request` already exists".

```zig
const Batch = union(enum) {
  newly_created: *Request,
  piggyback_onto_existing: struct { *Request, *Demux },
}; 
```

batiati replied:
I like the `union(enum)` idea.
We already have to check `demux == null` everywhere we use a batch, so let's use a proper `switch` statement instead.

The field's names can be shorter:
```Zig
const Batch = union(enum) {
  new: *Request,
  append: struct { *Request, *Demux },
};
```

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -196,50 +282,117 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+            };
+        }
+
+        /// After writing Event to batch.slice(), commit the batched
```
Incomplete sentence.

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -196,50 +282,117 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+                .callback = callback,
+            };
+
+            // Newly made Requests are populated and potentially pushed through VSR if no inflight.
```
Incomplete sentence.

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -568,9 +797,11 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
 
-        fn send_request_for_the_first_time(self: *Self, message: *Message) void {
-            assert(self.request_queue.head_ptr().?.message == message);
+        fn send_request_for_the_first_time(self: *Self) void {
```
I think I'd prefer if there was no `request_inflight` field and when needed we just check the head of the request queue. Otherwise there is an implicit (rather than explicit) contract in how `send_request_for_the_first_time` is used, which I think makes the code harder to follow.

kprotty replied:
Need to avoid batching into the head Request of the queue if it's currently inflight (message has already been sent to MessageBus). The implicit contract is indeed kind of a hack.

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -196,50 +282,117 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+                var it = self.request_queue.iterator_mutable();
+                while (it.next_ptr()) |request| {
+                    if (request == self.request_inflight) continue;
+                    if (request.message.header.command != .request) continue;
```
How is the command ever not `request`?

kprotty replied:
Good point: `.register` can be put into the queue, but if so it's always `request_inflight`.

sentientwaffle replied:
`register` is the  operation; the command is still `request` even then.

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -26,10 +28,83 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+            request: *Request,
+            demux: ?*Request.Demux, // Null iff batch_get() newly created `request`
+
+            pub fn slice(batch: Batch) []u8 {
```
Assert the request's message's command.

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -26,10 +28,83 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+                    return &pool.memory[pool.used];
+                }
+                return null;
+            }
```
Why use the "lazy fifo" pattern rather than an IOPS? :thinking:

batiati replied:
IOPS is limited to 2^6 elements 🙁

sentientwaffle replied:
Hmm, is that arbitrary? We could just bump that capacity `u6` up to a large enough int to fix whatever the expected capacity here is. Admittedly it introduces a bitset (which the current `DemuxPool` avoids), but it would also remove ~30 lines of code.

kprotty replied:
IOPS uses an array for the elements and `batch_logical_max` was so large (12MB worth of Demux nodes) that it would make the vsr.Client overflow the stack.

kprotty committed:
vsr: simplify some client batching assumptions

kprotty committed:
vsr: Demuxer type-safety and unit tests

batiati commented on src/testing/state_machine.zig:
```diff
@@ -3,14 +3,13 @@ const assert = std.debug.assert;
 pub fn StateMachineType(
     comptime Storage: type,
-    comptime config: constants.StateMachineConfig,
+    comptime config: @import("../constants.zig").StateMachineConfig,
```
Let's use the same pattern `const global_constants = @import("constants.zig");` seen in the real  StateMachine.zig code.

batiati commented on src/vsr/client.zig:
```diff
@@ -26,10 +29,97 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+                        },
+                    }
+                } else blk: {
+                    assert(batch.request.demux_queue.empty());
```
We can also assert the inverse `assert(!batch.request.demux_queue.empty())` when a `demux` is present.

Maybe we could assert if the `demux` is contained in the `request.demux_queue` (enclosed by `if (constants.verify)`)

batiati commented:
### Benchmark

I just created an [async benchmark](https://gist.github.com/batiati/8bd19479480ea0962d783bc794ff907c) running 4096 concurrent tasks (via thread pool), each one calling `create_transfers` with **only a single transfer**: 

- branch main, 10K transfers without client batching:
```
547 transfers per second with 4096 concurrent tasks.
create_transfers max p100 latency per transfer = 14_499ms.
total 10_000 transfers in 18_272ms.
```

- branch king-client-batching, 1M transfers:
```
165_837 transfers per second with 4096 concurrent tasks.
create_transfers max p100 latency per transfer = 93ms.
total 1_000_000 transfers in 6030ms.
```

The TPS numbers are impressively close to the benchmark with 8191 transfers per batch! Well done!
The throughput depends only on the application's capacity to submit enough requests to fill a batch, limited by `concurrency_max`.

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -433,22 +626,74 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+            inflight.demux_queue.count += 1;
+
+            switch (operation) {
+                inline else => |op| {
```
Let's not shorten "operation" to "op", since op numbers are a thing. Maybe `operation_comptime`?

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -433,22 +626,74 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+                        std.debug.panic("invalid batched op: {s}", .{
+                            inflight_operation.tag_name(StateMachine),
+                        });
+                    }
```
Can this error be hit somehow? I don't understand why this has a `if () { panic }` rather than a simple assertion, e.g. `assert(StateMachine.batch_logical_allowed.get(op));`.

Also it might be more clear if it was moved up to after we fall through `if (inflight.demux_queue.empty())`?

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -100,6 +100,48 @@ pub fn StateMachineType(
+                pub fn decode(self: *Demuxer, event_offset: u32, event_size: u32) []CreateResult {
+                    // Count all results from the start of our slice which match the Event range,
+                    // Updating the result.indexes to be relative to the Event range in the process.
+                    const demuxed = for (self.results, 0..) |*result, i| {
```
Could this implementation be more efficient if we remember the previous batch's offset within `results`, rather than re-iterating over all of `results` for each batch that we demux?

e.g. right now I think if each batch has exactly one event, then demultiplexing is `O(n²)`.

kprotty replied:
The `results` itself is re-sliced with the found batch offset in the `defer = self.results = self.results[demuxed..]` below. This achieves a similar effect IIUC?

sentientwaffle replied:
You're correct, thanks!

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -26,10 +29,97 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
-            // Null iff operation=register.
-            callback: ?Callback,
+
+            const Info = struct {
```
`Info` is a very vague/generic name, is there a more descriptive name we could use?

Or even better, could we get rid of it entirely?:
- Inline `user_data`/`callback` into `Demux`.
- Use a single-item `demux_queue` for events that cannot be batched.

And maybe this would reduce dimensionality in `on_reply`, too?
- More shared code between batching and non-batching operations.
- In `StateMachine`, implement trivial `Demuxer`s for non-batchable operations.

kprotty replied:
Demux should be removed when switching to a Context-based API so could this change be deferred until then?

batiati replied:
It's a small change that simplifies the code, so it's worth including now.
We can benefit from that when migrating to Context-based. Demux will be removed from the vsr. Client, but some equivalent mechanism may be inspired by the current way it works.

kprotty committed:
vsr: code reorg + handle new Message.Header types

kprotty committed:
vsr: fix Message.Header and u128 checksum usage

kprotty committed:
zig fmt

kprotty committed:
vsr: remove Request.Info & Demux all ops

In vsr.Client, the `Info` struct was too vague so it was inlined into Demux.
This means the initial Request is also represented as a Demux node.
Combining this with Demuxer now supporting all ops makes the demuxing
logic simpler to follow.

kprotty committed:
vsr: clarify/assert Demuxer logic

kprotty committed:
sm: handle events properly in testing Demuxer

kprotty committed:
sm: remove invalid demuxer assert

batiati commented on src/vsr/client.zig:
```diff
@@ -424,28 +632,63 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+                    // Push the inflight's demux to the front of the demux_queue to simplify handling.
+                    inflight.demux.next = inflight.demux_queue.out;
+                    inflight.demux_queue.out = &inflight.demux;
+                    inflight.demux_queue.count += 1;
```
Suggestion: Rename `Request.demux` to something like `Request.demux_queue_head` and add it immediately to the `demux_queue` when creating a new request.
Here, we can just call `demux_queue.pop()` without extra logic.

kprotty replied:
It would be self-referential and `request_queue.pop()` returns a (RLS) copy from its ring buffer, so `demux_queue_head` in the FIFO would need a fixup after that call.

batiati replied:
Sadly true 😅
Well, in this case we have the option of acquiring from the DemuxPool in either cases, what do you think? No queue head required!

I think we can't exceed the DemixPool’s capacity the way it is calculated.

kprotty replied:
It could introduce a failure/unreachable in `raw_request` if `get_message()` doesn't pair well with demux nodes, although  that's highly unlikely. `raw_request` is also only used by testing code (aof/cluster) which should already ensure message limits. Ill go ahead.

kprotty committed:
vsr: no intrusive Demux nodes

batiati commented on src/vsr/client.zig:
```diff
@@ -1,10 +1,13 @@
+noinline fn assert(cond: bool) void {
+    if (!cond) @panic("assert failed");
+}
+
```
I'm curious about that. 🤓
Why can't we use `const assert = std.debug.assert;` here?

kprotty replied:
Oops, forgot to remove debugging code. This was for stack trace on windows (`unreachable` didn't provide any)

batiati commented on src/vsr/client.zig:
```diff
@@ -414,10 +417,10 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
-            if (batch.demux == &batch.request.demux) {
-                assert(batch.request.demux_queue.empty());
+            // A Batch is newly made if its Request only contains its Demux.
+            // NOTE: This sets the `batch.request.message.request` number.
```
```zig
// NOTE: This sets the `batch.request.message.request` number.
```
I don't think this comment is up to date, the request number was already set during `batch_get`.

Also, we can swap the order of `self.register()` with the `header` initialization, so we don't need to use `.request = undefined` when initializing the header anymore.

batiati replied:
>```zig
> // NOTE: This sets the `batch.request.message.request` number.
>```
> I don't think this comment is up to date, the request number was already set during `batch_get`.

In fact, I DO think this comment was correct!

https://github.com/tigerbeetle/tigerbeetle/blob/74312b226bbb0921e006993c6c3298e0a8457bb7/src/vsr/client.zig#L382-L386

It raised my concern about calling `register()` during `batch_get()`.
I think `batch_get()` was supposed to behave like `message_get()`,  and `register()` should only be called by `batch_submit`.

kprotty replied:
`batch_get()` needs to call `register()` so that the latter's Request is pushed to the `request_queue` first. `batch_get()` pushes to the queue to obtain a `*Request` from the tail. Maybe it could have its own `Request`, populate that, and push to queue on `batch_submit()`? This would mean the following isn't batchable (Should we tighten doc-comment to make this not even a valid use-case?)
```zig
const a = batch_get(.batchable_op, max_events - 1);
const b = batch_get(.batchable_op, 1);
```

batiati replied:
I think the docs are clear enough that `batch_get` must be followed by `batch_submit` immediately, so it's not a valid use case (I can't think of an `assert` here, but it'd be much better than docs).

> he latter's Request is pushed to the `request_queue` first. `batch_get()` pushes to the queue to obtain a `*Request` from the tail. Maybe it could have its own Request, populate that, and push to queue on batch_submit()? 

I remember I asked this same question in the past, and I ended up deferring to enqueue the request to `batch_submit`, holding either a `Message` or a `Request` depending on the case.

https://github.com/tigerbeetle/tigerbeetle/blob/14b7afc2aa389394630d13545a38734e64271d5f/src/vsr/client.zig#L31-L46

But you're right, I was most concerned about keeping it with the same semantics of `message_get`, but I don't think it makes any difference with the intended usage.

Just adjust the outdated comment, and we're good to go :+1:

batiati commented on src/vsr/client.zig:
```diff
@@ -382,18 +389,13 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
             });
 
             const request = self.request_queue.tail_ptr().?;
+            request.demux_queue.push(demux);
```
Assert that `demux_queue.count == 0` here.
Also, we can assert that `demux_queue.count >1` above when appending a demux into an existing request.

batiati reviewed (commented):
Great job @kprotty 🥳
There are just some minor comments to be addressed here, otherwise LGTM.
I think @sentientwaffle wants to give it one more pass before approval.

kprotty committed:
vsr: remove testing code + demux_queue asserts

kprotty committed:
vsr: loosen demux_queue assert when batching

sentientwaffle commented on src/benchmark.zig:
```diff
@@ -397,22 +397,17 @@ const Benchmark = struct {
-        );
+        const event_size = switch (operation) {
+            inline else => |op| @divExact(payload.len, @sizeOf(StateMachine.Event(op))),
+        };
```
Shouldn't this be named `event_count`?

sentientwaffle commented on src/clients/c/tb_client/echo_client.zig:
```diff
@@ -62,32 +68,70 @@ pub fn EchoClient(comptime StateMachine_: type, comptime MessageBus: type) type
+                .event_count = @intCast(event_count),
+                .demux = demux,
+            };
+        }
```
So to check my understanding, the echo client's `batch_get`/`batch_submit` don't actually do any batching, right? i.e. each "batch_get" ends up in a separate request message no matter what.

But then why does `Batch` have a `demux` property at all?

sentientwaffle commented on src/clients/c/tb_client/echo_client.zig:
```diff
@@ -62,32 +68,70 @@ pub fn EchoClient(comptime StateMachine_: type, comptime MessageBus: type) type
+            event_count: usize,
+        ) Self.BatchError!Batch {
+            const body_size = switch (operation) {
+                inline else => |op| @sizeOf(Self.StateMachine.Event(op)) * event_count,
```
```suggestion
                inline else => |operation| @sizeOf(Self.StateMachine.Event(operation)) * event_count,
```

"op" typically refers to an op number, not the state machine operation type.

sentientwaffle commented on src/vsr/client.zig:
```diff
@@ -196,52 +296,131 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
+            if (StateMachine.batch_logical_allowed.get(operation)) {
+                var it = self.request_queue.iterator_mutable();
+                while (it.next_ptr()) |request| {
+                    if (request == self.request_inflight) continue;
```
Re: https://github.com/tigerbeetle/tigerbeetle/pull/1257#discussion_r1384075559:

> Need to avoid batching into the head Request of the queue if it's currently inflight (message has already been sent to MessageBus). The implicit contract is indeed kind of a hack.

Could we accomplish this (minus the hack) by just calling `_ = it.next_ptr();` immediately before this loop starts? (With a comment to explain.)

kprotty committed:
vsr: implicit request_inflight from request_queue.head

kprotty committed:
vsr/client: simplify EchoClient batching

kprotty committed:
benchmark: rename event_size to event_count

batiati reviewed (approved):
Awesome work @kprotty!
